presidential-speeches-rag/presidential-speeches-rag.ipynb

```diff
@@ -0,0 +1,657 @@
+       "2  In a wide ranging speech, President Washington...   \n",
+       "3  Washington focuses on commerce in his second a...   \n",
+       "4  The President reassures the Seneca Nation that...   \n",
+       "\n",
+       "                                          Transcript   \n",
+       "0  Fellow Citizens of the Senate and the House of...  \\\n",
+       "1  Whereas it is the duty of all Nations to ackno...   \n",
+       "2  Fellow Citizens of the Senate and House of Rep...   \n",
+       "3  Fellow citizens of the Senate and House of Rep...   \n",
+       "4  I the President of the United States, by my ow...   \n",
+       "\n",
+       "                                                 URL  \n",
+       "0  https://millercenter.org/the-presidency/presid...  \n",
+       "1  https://millercenter.org/the-presidency/presid...  \n",
+       "2  https://millercenter.org/the-presidency/presid...  \n",
+       "3  https://millercenter.org/the-presidency/presid...  \n",
+       "4  https://millercenter.org/the-presidency/presid...  "
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "presidential_speeches_df = pd.read_csv('presidential_speeches.csv')\n",
+    "presidential_speeches_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a9aaabfb-c34d-40f1-a90f-f448a9051130",
+   "metadata": {},
+   "source": [
+    "To get a better idea of the steps involved in building a RAG system, let's focus on a single speech to start. In honor of his [upcoming Netflix series](https://www.netflix.com/tudum/articles/death-by-lightning-tv-series-adaptation) and his distinction of being the only president to [contribute an original proof of the Pythagorean Theorem](https://maa.org/press/periodicals/convergence/mathematical-treasure-james-a-garfields-proof-of-the-pythagorean-theorem), we'll use James Garfield's Inaugural Address:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "39439748-8652-415d-a5e7-0f421a6ae30a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "garfield_inaugural = presidential_speeches_df.iloc[309].Transcript\n",
+    "#display(HTML(garfield_inaugural)) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8e1a9811-2fd6-4c99-ba11-a9df2df33ec0",
+   "metadata": {},
+   "source": [
+    "A challenge with prompting LLMs can be running into limits with their context window. While this speech is not extremely long and would actually in Mixtral's context window, it is not always great practice to use way more of the context window than you need, so when using RAG we want to split up the text to provide only relevant parts of it to the LLM. To do so, we first need to ```tokenize``` the transcript. We'll use the Mixtral 8x7b tokenzier with the transformers AutoTokenizer class for this - this will show the number of tokens the Mixtral 8x7b model counts in Garfield's Inaugural Address:"
```

Comment:
```suggestion
    "A challenge with prompting LLMs can be running into limits with their context window. While this speech is not extremely long and would actually fit in Mixtral's context window, it is not always great practice to use way more of the context window than you need, so when using RAG we want to split up the text to provide only relevant parts of it to the LLM. To do so, we first need to ```tokenize``` the transcript. We'll use the Mixtral 8x7b tokenzier with the transformers AutoTokenizer class for this - this will show the number of tokens the Mixtral 8x7b model counts in Garfield's Inaugural Address:"
```